# Data 200: Database Systems and Data Management for Data Analytics


# Homework 6: Transformations in Pandas

**Dickinson College**<br/>
**Spring 2022**<br/>
**Instructor:** Dick Forrester<br/>
**Author:** Dick Forrester<br>
<font color='red'>**Due Date and Time:** 11:59pm on Thursday, 2/24/2022 </font>
---
Enter your name in the markdown cell below.

# Name:

In [1]:
## RUN THIS CELL TO GET THE RIGHT FORMATTING AND TO LOAD NumPy
import requests
import numpy as np
from IPython.core.display import HTML
styles = requests.get("https://raw.githubusercontent.com/Harvard-IACS/2018-CS109A/master/content/styles/cs109.css").text
HTML(styles)

# Tasks

- Review pages 133-151 in the Course Notes.
- Complete the **Advanced Merging and Concatenating** and **Merging Ordered and Time-Series Data** chapters of the **Joining Data with Pandas** course on DataCamp
- E-mail me your completed Jupyter notebook.

# Exercises

This homework will give you some practice using *melt* and *pivot* in Pandas.  The data for this assignment is contained in the `hw06data` folder.

<div class="exercise"><b>Exercise 1: Loading and Examining Air Quality Data</b></div> 

The `airquality.csv` file in the folder `hw06data` contains some data related to air quality.

In the code cell below, write Python code to do the following.<br>

1. Import the Pandas library.
1. Read in the file `airquality.csv` and assign it to the data frame `airquality`.
1. Print `airquailty.head()` and examine the columns and type of data contained in the data frame.
1. Print `airquality.shape` to determine the number of rows and columns.

The output from my solution is:<br>
<code>
   Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5
</code><code>
(153, 6)
</code>

In [2]:
import pandas as pd
airquality = pd.read_csv('hw06data/airquality.csv')
print(airquality.head())
print()
print(airquality.shape)

   Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

(153, 6)


<div class="exercise"><b>Exercise 2: Melting Air Quality Data</b></div> 

Note that `airquality` is a *tidy* data frame because all the variables (`Ozone`, `Solar.R`, `Wind`, and `Temp`) each have their own column and each row represent a different observation. 

Suppose we wanted these variables to be in rows instead, which would of course make the data *untidy*! To accomplish this we can use the `melt()` function, which has two main parameters: `id_vars` and `value_vars`. The `id_vars` represent the columns of the data you do not want to melt (i.e., keep it in its current shape), while the `value_vars` represent the columns you do wish to melt into rows. By default, if no `value_vars` are provided, all columns not set in the `id_vars` will be melted. This could save a bit of typing, depending on the number of columns that need to be melted.

In the code cell below, write Python code to do the following.<br>

1. Use `melt()` to melt the `Ozone`, `Solar.R`, `Wind`, and `Temp` columns into rows and assign it to the data frame `airquality_melt`.
1. Print the first five rows of `airquality_melt`.
1. Print `airquality_melt.shape`

The output from my solution is:<br>
<code>
   Month  Day variable  value
0      5    1    Ozone   41.0
1      5    2    Ozone   36.0
2      5    3    Ozone   12.0
3      5    4    Ozone   18.0
4      5    5    Ozone    NaN
</code><code>
(612, 4)
</code>

In [3]:
airquality_melt = airquality.melt(id_vars = ['Month','Day'], \
                                 value_vars = ['Ozone','Solar.R','Wind','Temp'])
print(airquality_melt.head())
print()
print(airquality_melt.shape)

   Month  Day variable  value
0      5    1    Ozone   41.0
1      5    2    Ozone   36.0
2      5    3    Ozone   12.0
3      5    4    Ozone   18.0
4      5    5    Ozone    NaN

(612, 4)


It's important to note that the `airquality_melt` data frame is no longer tidy.  Furthermore, note that the third column in `airquality_melt` is generically called `variable`, while the last column is generically called `value`.

<div class="exercise"><b>Exercise 3: Melting Air Quality Data Part 2</b></div> 

When melting data frames, it would be better to have column names more meaningful than `variable` and `value`. The default names may work in certain situations, but it's best to always have data that is self explanatory. You can rename the `variable` column by specifying an argument to the `var_name` parameter, and the `value` column by specifying an argument to the `value_name` parameter.

In the code cell below, write Python code to do the following.<br>

1. Update your call to `melt()` from the previous exercise so that is specifies `var_name='measurement'` and `value_name='reading'` as additional parameters (which provide appropriate column names), once again assign it to the data frame `airquality_melt`.
1. Print the first five rows of `airquality_melt`.
1. Print `airquality_melt.shape`

The output from my solution is:<br>
<code>
   Month  Day measurement  reading
0      5    1       Ozone     41.0
1      5    2       Ozone     36.0
2      5    3       Ozone     12.0
3      5    4       Ozone     18.0
4      5    5       Ozone      NaN
</code><code>
(612, 4)
</code>

In [4]:
airquality_melt = airquality.melt(id_vars = ['Month','Day'], \
                                 value_vars = ['Ozone','Solar.R','Wind','Temp'], \
                                 var_name = 'measurement', value_name = 'reading')
print(airquality_melt.head())
print()
print(airquality_melt.shape)

   Month  Day measurement  reading
0      5    1       Ozone     41.0
1      5    2       Ozone     36.0
2      5    3       Ozone     12.0
3      5    4       Ozone     18.0
4      5    5       Ozone      NaN

(612, 4)


The DataFrame is more informative now because of the appropriate column names.

<div class="exercise"><b>Exercise 4: Pivoting Melting Air Quality Data</b></div> 

**Pivoting** data is the opposite of melting it. Remember the tidy form that the `airquality` data frame was in before you melted it? You'll now begin pivoting it back into that form using the `pivot_table()` method!

While melting takes a set of columns and turns it into a single column, pivoting will create a new column for each unique value in a specified column.  The `pivot_table()` function has an `index` parameter which you can use to specify the columns that you don't want pivoted: It is similar to the `id_vars` parameter of `pd.melt()`. Two other parameters that you have to specify are `columns` (the name of the column you want to pivot), and `values` (the values to be used when the column is pivoted).

In the code cell below, write Python code to do the following.<br>

1. Using `pd.pivot_table()`, *pivot* on the `measurement` column, *don't* pivot on the `Month` and `Day` columns, and use `reading` as the values to use when the column is pivoted.  Call the resulting data frame `airquality_pivot`.
1. Print the first five rows of `airquality_pivot`.
1. Print `airquality_pivot.shape`

The output from my solution is:<br>
<code>
measurement  Ozone  Solar.R  Temp  Wind
Month Day                              
5     1       41.0    190.0  67.0   7.4
      2       36.0    118.0  72.0   8.0
      3       12.0    149.0  74.0  12.6
      4       18.0    313.0  62.0  11.5
      5        NaN      NaN  56.0  14.3
</code><code>
(153, 4)
</code>

In [6]:
airquality_pivot = airquality_melt.pivot_table(index = ["Month","Day"], columns = "measurement", values = "reading")
print(airquality_pivot.head(5))
print(airquality_pivot.shape)

measurement  Ozone  Solar.R  Temp  Wind
Month Day                              
5     1       41.0    190.0  67.0   7.4
      2       36.0    118.0  72.0   8.0
      3       12.0    149.0  74.0  12.6
      4       18.0    313.0  62.0  11.5
      5        NaN      NaN  56.0  14.3
(153, 4)


Notice that after pivoting `airquality_melt` we obtained a data frame with a *hierarchical index* (also known as a `MultiIndex`).  In essence, they allow you to group columns or rows by another variable - in this case, by `Month` as well as `Day`.

If we wanted to get back to our original data frame before the melt, that is, we wanted to remove the hierarchical index, we can use the `reset_index()` method.  Run the code cell below and examine the results.

In [7]:
# Reset the index of airquality_pivot: airquality_pivot_reset
airquality_pivot_reset = airquality_pivot.reset_index()

# Print the head of airquality_pivot_reset
print(airquality_pivot_reset.head())

measurement  Month  Day  Ozone  Solar.R  Temp  Wind
0                5    1   41.0    190.0  67.0   7.4
1                5    2   36.0    118.0  72.0   8.0
2                5    3   12.0    149.0  74.0  12.6
3                5    4   18.0    313.0  62.0  11.5
4                5    5    NaN      NaN  56.0  14.3


## Handling Missing Data
Before we move on to our next data set, let's once again examine the original `airquality` data frame.  Run the code cell below.

In [8]:
print(airquality.head())
print()
print(airquality.info())

   Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Ozone    116 non-null    float64
 1   Solar.R  146 non-null    float64
 2   Wind     153 non-null    float64
 3   Temp     153 non-null    int64  
 4   Month    153 non-null    int64  
 5   Day      153 non-null    int64  
dtypes: float64(3), int64(3)
memory usage: 7.3 KB
None


First, notice that in row 4 there is a `NaN` in both the `Ozone` and `Solar.R` columns, indicating  *mising values*.  Furthermore, we can see from the output of the `info()` method that there are `153 entires` (rows), but for the `Ozone` column there are only `116 non-null` values.  This implies that there are 153-116 = 37 missing values.  **Please make sure you see where this is indicated in the output**.

Missing values are common in dealing with real-world problems, especially when the data is aggregated over long time stretches. In statistics, **imputation** is the process of replacing missing data with substituted values. It's important to deal with missing data because certain calculations cannot handle missing values while some calculations will, by default, skip over any missing values.
Also, understanding how much missing data you have, and thinking about where it comes from is crucial to making unbiased interpretations of data.

Run the code cell below, which is the `fillna()` method to replace all the `NaN` values in the `Ozone` column with the mean of the `Ozone` values. You can read more about the `fillna()` function [here](https://www.geeksforgeeks.org/python-pandas-dataframe-fillna-to-replace-null-values-in-dataframe/).  **Make sure you understand the code before moving on!**


In [9]:
# Calculate the mean of the Ozone column: oz_mean
oz_mean = airquality.Ozone.mean()

# Replace all the missing values in the Ozone column with the mean
airquality['Ozone'] = airquality.Ozone.fillna(oz_mean)

# Print the info of airquality
print(airquality.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Ozone    153 non-null    float64
 1   Solar.R  146 non-null    float64
 2   Wind     153 non-null    float64
 3   Temp     153 non-null    int64  
 4   Month    153 non-null    int64  
 5   Day      153 non-null    int64  
dtypes: float64(3), int64(3)
memory usage: 7.3 KB
None


Notice that now there are `153 non-null` values in the `Ozone` column, matching the total number of rows.  In particular, all the `NaN` values have been replaced with the mean of the `Ozone` values.

<div class="exercise"><b>Exercise 5: Loading and Examining Tuberculosis Data</b></div> 

The `tb.csv` file in the folder `hw06data` consisting of case counts of tuberculosis by country, year, gender, and age group.

In the code cell below, write Python code to do the following.<br>

1. Read in the file `tb.csv` and assign it to the data frame `tb`.
1. Print `tb.head()` and examine the columns and type of data contained in the data frame.

The output from my solution is:<br>
<code>
  country  year  m014  m1524  m2534  f014  f1524  f2534
0      AD  2000   0.0    0.0    1.0   NaN    NaN    NaN
1      AE  2000   2.0    4.0    4.0   3.0   16.0    1.0
2      AF  2000  52.0  228.0  183.0  93.0  414.0  565.0
3      AG  2000   0.0    0.0    0.0   1.0    1.0    1.0
4      AL  2000   2.0   19.0   21.0   3.0   11.0   10.0
</code>

In [10]:
tb = pd.read_csv('hw06data/tb.csv')
print(tb.head())

  country  year  m014  m1524  m2534  f014  f1524  f2534
0      AD  2000   0.0    0.0    1.0   NaN    NaN    NaN
1      AE  2000   2.0    4.0    4.0   3.0   16.0    1.0
2      AF  2000  52.0  228.0  183.0  93.0  414.0  565.0
3      AG  2000   0.0    0.0    0.0   1.0    1.0    1.0
4      AL  2000   2.0   19.0   21.0   3.0   11.0   10.0


<div class="exercise"><b>Exercise 6: Tidying the Tuberculosis Data</b></div> 

Note that `m014` represents males aged 0-14, `f014` represents females aged 0-14, etc.  In this exercise, you're going to *tidy* the `m014-f2534` columns by creating a `gender` and `age_group` column.  

In the code cell below, write Python code to do the following.<br>

1. Use `melt()` to melt the `m014`,  `m1524`,  `m2534`,  `f014`,  `f1524`,  and `f2534` columns into rows and assign it to the data frame `tb_melt`.
1. Create a column called `gender` that extracts the first letter of the `variable` column that resulted from the melt (containing the strings `m014`, `m1524`, etc).  Thus, `gender` should be a column containing only the letters `m` or `f`.
1. Create a column called `age_group` that extracts the remaining letters of the `variable` column that resulted from the melt (containing the strings `m014`, `m1524`, etc).  Thus, `age_group` should be a column containing only values such as `014`,  `1524`,  or `2534`.
1. Finally, drop the `variable` column from the `tb_melt` data frame using the `drop()` method (since it's not needed any more)--you may want to review page 120 of the Course Notes to remember how to do this.
1. Print `tb_melt.head()`.
1. Print `tb_melt.tail()`.

The output from my solution is:<br>
<code>
  country  year  value gender age_group
0      AD  2000    0.0      m       014
1      AE  2000    2.0      m       014
2      AF  2000   52.0      m       014
3      AG  2000    0.0      m       014
4      AL  2000    2.0      m       014
</code><code>
     country  year   value gender age_group
1201      YE  2000   627.0      f      2534
1202      YU  2000     NaN      f      2534
1203      ZA  2000  1716.0      f      2534
1204      ZM  2000  1118.0      f      2534
1205      ZW  2000     NaN      f      2534
</code>

In [11]:
tb_melt = tb.melt(id_vars=['country','year'], value_vars = ['m014', 'm1524', 'm2534', 'f014', 'f1524', 'f2534'])
tb_melt['gender'] = tb_melt['variable'].str[0]
tb_melt['age_group'] = tb_melt['variable'].str[1:]
tb_melt.drop('variable', axis = 1, inplace = True)
print(tb_melt.head())
print(tb_melt.tail())

  country  year  value gender age_group
0      AD  2000    0.0      m       014
1      AE  2000    2.0      m       014
2      AF  2000   52.0      m       014
3      AG  2000    0.0      m       014
4      AL  2000    2.0      m       014
     country  year   value gender age_group
1201      YE  2000   627.0      f      2534
1202      YU  2000     NaN      f      2534
1203      ZA  2000  1716.0      f      2534
1204      ZM  2000  1118.0      f      2534
1205      ZW  2000     NaN      f      2534


This data is now tidy and hopefully would be easier to analyze!